open_esplibs/libnet80211/ieee80211_hostap.c

```diff
@@ -0,0 +1,34 @@
+/* Recreated Espressif libnet80211 ieee80211_hostap.o contents.
+
+   Copyright (C) 2015 Espressif Systems. Derived from MIT Licensed SDK libraries.
+   BSD Licensed as described in the file LICENSE
+*/
+#include "open_esplibs.h"
+#if OPEN_LIBNET80211_HOSTAP
+// The contents of this file are only built if OPEN_LIBNET80211_HOSTAP is set to true
+
+#include "sdk_internal.h"
+
+void sdk_ieee80211_send_mgmt(void *, int, int);
+void sdk_cnx_node_leave(struct sdk_g_ic_netif_info *netif, struct sdk_netif_conninfo *conn);
+
+void sdk_hostap_handle_timer(struct sdk_netif_conninfo *cnx_node)
+{
+    uint32_t count1 = *(uint32_t *)(((void *)cnx_node) + 0x108);
+    uint32_t count2 = TIMER_FRC2.COUNT;
+
+    if ((count2 - count1) > 93599688U) {
```

Comment:
@ourairquality You seems to have optimised the code here :)
I looked into the disassembly of that part and don't quite understand the intended logic:
```
            l32i         a0, a12, 0x108 /* [arg0+0x108] */     #  52b: 022c42
// a0 = cnx_node->0x108
            l32i         a2, a2, 36 /* [0x60000624] */         #  52e: 2892
// a2 = FRC2.COUNT
            bltu         a2, a0, .Lbr021                       #  530: 07320e
// if (a2 >= a0) {
            movi         a3, 312 /* 0x00000138 */              #  533: 32a138
            sub          a2, a2, a0                            #  536: 0022c0
            call0        __udivsi3                             #  539: 01f3ffc0
                                                               #  53d: 0000
            j            .Lbr022                               #  53f: 460400
// if (a2 < a0) {
.Lbr021:    movi         a3, -1 /* 0xffffffff */               #  542: 7cf3
            xor          a3, a0, a3                            #  544: 303030
            add          a2, a2, a3                            #  547: 3a22
            addi         a2, a2, 1                             #  549: 1b22
            movi         a3, 312 /* 0x00000138 */              #  54b: 32a138
            call0        __udivsi3                             #  54e: 01efffc0

.Lbr022:    movi         a4, 0x000493df                        #  554: 41e9ff
            bgeu         a4, a2, .Lbr023                       #  557: 27b43b

```
So, the the logic seems to be the following:
```
if (curr_time < timestamp) {
    t = (curr_time + (~timestamp) + 1) / 312;
} else {
    t = (curr_time - timestamp) / 312;
}

if ((300000 - 1) < t) {  // (300000-1) == 0x000493df
```
But the expression `(~timestamp) + 1` equals `-timestamp`. So, both branches basically do the same.
@ourairquality Do you have any clues what might be the reason behind all this?

BTW, great job on reverse engineering those functions!